### DISCUSSION TITLE
feat: canvas

### DISCUSSION THREAD

**Discussion by luantak** [open-webui/open-webui#6684] [General]
Similar to https://openai.com/index/introducing-canvas/

The model will need to be instructed to output diffs, in order for this to work.

Generated code should also be able to be run directly.

**Comment by ipatman** [comment]
Maybe this could be integrated
https://github.com/langchain-ai/open-canvas

**Reply by weareturnone** [reply]
Would be great

**Reply by sclausen** [reply]
fyi: Unfortunately, this is not being able to run offline, since the dependency of several external services, as well as the need to obtain an OpenAI, Anthropic and LangSmith API key.

**Comment by niderhoff** [comment]
> Note taking feature (similar to Notion/Obsidian), to be used alongside the current chat interface, will be worked on from the ground up and is on the roadmap. Stay tuned!

https://github.com/open-webui/open-webui/discussions/6009

**Comment by healdgar** [comment]
Gemini created a Canvas feature, but it lacks redlines, which to me defeats the whole purpose of a Canvas... If you can't see the diff/redline, you can't see what changes the model made and the model make make unauthorized edits.  I hope whatever implementation of this feature we make for Open Webui has diff/redlines.

**Comment by raythurnvoid** [comment]
We recently shipped this in our product, i can say it's a shit show due to the lack of good rich text editors packages due to poor native support in general.

Tiptap is hard, takes time, you always end up fighting with it.

Blocknote is built on top of tiptap (opencanvas uses that), it gives you something more abstract, but even the features they implemented are somewhat buggy. I mean they did a terrific job but the problem is just that it's hard and takes time and money to make good stuff.

**Comment by YousefED** [comment]
We've explored a prototype connecting OpenWebUI with the [BlockNote rich text editor](https://www.blocknotejs.org/) to create a Canvas like experience

Here's an example of what it could look like:

https://github.com/user-attachments/assets/904ccd63-696b-41bc-97c9-1171c38ec4a0

Would be open to partner up and develop this further!

**Comment by weareturnone** [comment]
Awesome!!

**Comment by Buckler89** [comment]
Any update on that?
